function-service-impl/src/main/java/com/future/function/service/impl/feature/communication/chatroom/MessageStatusServiceImpl.java

```diff
@@ -1,140 +1,158 @@
+import com.future.function.model.entity.feature.core.User;
+import com.future.function.repository.feature.communication.chatting.MessageStatusRepository;
+import com.future.function.service.api.feature.communication.chatroom.ChatroomService;
+import com.future.function.service.api.feature.communication.chatroom.MessageService;
+import com.future.function.service.api.feature.communication.chatroom.MessageStatusService;
+import com.future.function.service.api.feature.core.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.redis.core.RedisTemplate;
+import org.springframework.data.redis.core.SetOperations;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class MessageStatusServiceImpl implements MessageStatusService {
+
+  private final UserService userService;
+
+  private final MessageStatusRepository messageStatusRepository;
+
+  private final ChatroomService chatroomService;
+
+  private final MessageService messageService;
+
+  private final SetOperations<String, Object> redisSetOperations;
+
+
+  @Autowired
+  public MessageStatusServiceImpl(
+    UserService userService, MessageStatusRepository messageStatusRepository,
+    ChatroomService chatroomService, MessageService messageService, RedisTemplate<String, Object> redisTemplate
+  ) {
+
+    this.userService = userService;
+    this.chatroomService = chatroomService;
+    this.messageStatusRepository = messageStatusRepository;
+    this.messageService = messageService;
+    this.redisSetOperations = redisTemplate.opsForSet();
+
+  }
+
+  @Override
+  public List<MessageStatus> getUnseenMessageStatus(
+    String chatroomId, String userId
+  ) {
+
+    User user = userService.getUser(userId);
+    Chatroom chatroom = chatroomService.getChatroom(chatroomId, userId);
+    return messageStatusRepository.findAllByChatroomAndMemberAndSeenIsFalseOrderByCreatedAtDesc(
+      chatroom, user);
+  }
+
+  @Override
+  public List<MessageStatus> getUnseenMessageStatusBeforeTimestamp(
+    String chatroomId, String userId, Long timestamp
+  ) {
+
+    User user = userService.getUser(userId);
+    Chatroom chatroom = chatroomService.getChatroom(chatroomId, userId);
+    return messageStatusRepository.findAllByChatroomAndMemberAndCreatedAtLessThanEqualAndSeenIsFalseOrderByCreatedAtDesc(
+      chatroom, user, timestamp);
+  }
+
+  @Override
+  public boolean getSeenStatus(String chatroomId, String userId) {
+
+    return this.getUnseenMessageStatus(chatroomId, userId)
+             .size() == 0;
+  }
+
+  @Override
+  public MessageStatus createMessageStatus(
+    MessageStatus messageStatus, String userId
+  ) {
+
+    return Optional.of(messageStatus)
+      .map(msgStatus -> this.setChatroom(msgStatus, userId))
+      .map(this::setMember)
+      .map(this::setMessage)
+      .map(messageStatusRepository::save)
+      .orElseThrow(UnsupportedOperationException::new);
```

Comment:
consider giving a message to this as well

Reply:
it is not a new code anyway, but i don't know somehow it showed as addition